grammar.js

```diff
@@ -944,6 +951,12 @@ module.exports = grammar({
       ')'
     )),
 
+    opt_arguments: $ => prec(PREC.CALL, seq(
+      '?.(',
```

Comment:
Can you split the `?.` into a separate token?

Reply:
I think @mjambon mentioned that but this led to parsing errors when applying it on a concrete file.

Reply:
Exactly right. Using `seq('?.', '(', ...)` did not parse.

Reply:
What specifically was failing to parse? I'm trying this locally, with the following added unit test, and everything seems ok:

```
============================================
Optional function calls
============================================

a[b]?.(c);
d.e?.(f);

---

(program
  (expression_statement
    (call_expression
      (subscript_expression (identifier) (identifier))
      (opt_arguments (identifier))))
  (expression_statement
    (call_expression
      (member_expression (identifier) (property_identifier))
      (opt_arguments (identifier)))))

```

Reply:
This is my grammar diff:

```diff
--- a/grammar.js
+++ b/grammar.js
@@ -685,7 +685,8 @@ module.exports = grammar({
         $.super,
         alias($._reserved_identifier, $.identifier)
       )),
-      choice('[', seq('?.', '[')),
+      optional('?.'),
+      '[',
       field('index', $._expressions),
       ']'
     )),
@@ -952,7 +953,8 @@ module.exports = grammar({
     )),
 
     opt_arguments: $ => prec(PREC.CALL, seq(
-      '?.(',
+      '?.',
+      '(',
       commaSep(optional(choice($._expression, $.spread_element))),
       ')'
     )),
```

Reply:
It's the simpler case `a?.(b)` that fails:

```
============================================
Optional function calls
============================================

a?.(b);
a[b]?.(c);
d.e?.(f);

---
(program
  (expression_statement
    (call_expression (identifier) (arguments (identifier))))
  (expression_statement
    (call_expression
      (subscript_expression (identifier) (identifier))
      (opt_arguments (identifier))))
  (expression_statement
    (call_expression
      (member_expression (identifier) (property_identifier))
      (opt_arguments (identifier)))))
```

Adding all the changes to the branch, including the failing test case.

Reply:
Hi @maxbrunsfeld 

Using your `grammar.js` diff, and a test file of:

```js
x?.(y);

x ?. (y);

x?.y;

x?.[0];

new function bob (a, b) {} ?.(1, 2);
```

I got:

```
➜  tree-sitter-javascript git:(optional-chains) ✗ npx tree-sitter parse opt.js
(program [0, 0] - [9, 0]
  (expression_statement [0, 0] - [0, 7]
    (call_expression [0, 0] - [0, 6]
      function: (identifier [0, 0] - [0, 1])
      (ERROR [0, 1] - [0, 3])
      arguments: (arguments [0, 3] - [0, 6]
        (identifier [0, 4] - [0, 5]))))
  (expression_statement [2, 0] - [2, 9]
    (call_expression [2, 0] - [2, 8]
      function: (identifier [2, 0] - [2, 1])
      (ERROR [2, 2] - [2, 4])
      arguments: (arguments [2, 5] - [2, 8]
        (identifier [2, 6] - [2, 7]))))
  (expression_statement [4, 0] - [4, 5]
    (member_expression [4, 0] - [4, 4]
      object: (identifier [4, 0] - [4, 1])
      property: (property_identifier [4, 3] - [4, 4])))
  (expression_statement [6, 0] - [6, 7]
    (subscript_expression [6, 0] - [6, 6]
      object: (identifier [6, 0] - [6, 1])
      index: (number [6, 4] - [6, 5])))
  (expression_statement [8, 0] - [8, 36]
    (call_expression [8, 0] - [8, 35]
      function: (new_expression [8, 0] - [8, 26]
        constructor: (function [8, 4] - [8, 26]
          name: (identifier [8, 13] - [8, 16])
          parameters: (formal_parameters [8, 17] - [8, 23]
            (identifier [8, 18] - [8, 19])
            (identifier [8, 21] - [8, 22]))
          body: (statement_block [8, 24] - [8, 26])))
      arguments: (opt_arguments [8, 27] - [8, 35]
        (number [8, 30] - [8, 31])
        (number [8, 33] - [8, 34])))))
opt.js	0 ms	(ERROR [0, 1] - [0, 3])```

Reply:
Note that `tree-sitter` seems to be preferring the `$.arguments` rule instead of `$.opt_arguments` here. I played with various `prec` settings, but that didn't help me out any.

Reply:
@maxbrunsfeld now if you run the tests, you should see an ERROR node where the `?.` is:
```
1 failure:

expected / actual

  1. Optional function calls:
    (program (expression_statement (call_expression (identifier) (ERROR) (arguments (identifier)))) (expression_statement (call_expression (subscript_expression (identifier) (identifier)) (opt_arguments (identifier)))) (expression_statement (call_expression (member_expression (identifier) (property_identifier)) (opt_arguments (identifier))))) 
```

Reply:
Looks like @mjambon beat me to it, but you can reference https://github.com/returntocorp/tree-sitter-javascript/commit/7633d910c91a4f978e5de939682954dcf90a606f for my test case

Reply:
(thanks for your help here @maxbrunsfeld , definitely new to tree-sitter here and still learning the ropes!)